Deleting schmea label for Schema with a lot of datasets throws `OptimisticLockException`

When trying to delete a `Label` from a `Schema` where the schema is used by 1000's of datasets, the "save" button appears to freeze, a single core on my machine is pinned at 100% utilization;

![Screenshot from 2023-10-30 07-28-54](https://github.com/Hyperfoil/Horreum/assets/959822/1c5eb580-1ccf-4589-ab69-90d62aae2d1e)


Eventually the following exception is thrown.

```
2023-10-30 07:29:25,697 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-7) HTTP Request to /api/schema/1/labels/2799 failed, error id: 50a73092-1478-473b-bb59-f154c232d713-1: jakarta.persistence.OptimisticLockException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [io.hyperfoil.tools.horreum.entity.data.LabelDAO#2799]
	at org.hibernate.internal.ExceptionConverterImpl.wrapStaleStateException(ExceptionConverterImpl.java:204)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:93)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:162)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:168)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1416)
	at org.hibernate.internal.SessionImpl.flush(SessionImpl.java:1398)
	at org.hibernate.query.sql.internal.NativeQueryImpl.prepareForExecution(NativeQueryImpl.java:599)
	at org.hibernate.query.spi.AbstractSelectionQuery.beforeQuery(AbstractSelectionQuery.java:391)
	at org.hibernate.query.spi.AbstractSelectionQuery.list(AbstractSelectionQuery.java:364)
	at org.hibernate.query.Query.getResultList(Query.java:119)
	at io.hyperfoil.tools.horreum.svc.SchemaServiceImpl.emitLabelChanged(SchemaServiceImpl.java:704)
	at io.hyperfoil.tools.horreum.svc.SchemaServiceImpl.doUpdateLabelForDelete(SchemaServiceImpl.java:478)
	at io.hyperfoil.tools.horreum.svc.SchemaServiceImpl.deleteLabel(SchemaServiceImpl.java:747)
	at io.hyperfoil.tools.horreum.svc.SchemaServiceImpl_Subclass.deleteLabel$$superforward(Unknown Source)
	at io.hyperfoil.tools.horreum.svc.SchemaServiceImpl_Subclass$$function$$16.apply(Unknown Source)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:73)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext$NextAroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:97)
	at io.hyperfoil.tools.horreum.server.RolesInterceptor.intercept(RolesInterceptor.java:70)
	at io.hyperfoil.tools.horreum.server.RolesInterceptor_Bean.intercept(Unknown Source)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:70)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:62)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:136)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:107)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.doIntercept(TransactionalInterceptorRequired.java:38)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.intercept(TransactionalInterceptorBase.java:61)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.intercept(TransactionalInterceptorRequired.java:32)
	at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired_Bean.intercept(Unknown Source)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:30)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:27)
	at io.hyperfoil.tools.horreum.svc.SchemaServiceImpl_Subclass.deleteLabel(Unknown Source)
	at io.hyperfoil.tools.horreum.svc.SchemaServiceImpl_ClientProxy.deleteLabel(Unknown Source)
	at io.hyperfoil.tools.horreum.api.services.SchemaService$quarkusrestinvoker$deleteLabel_9290527456a709c6d06ca9d108a30e9ec091665b.invoke(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.hibernate.StaleObjectStateException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [io.hyperfoil.tools.horreum.entity.data.LabelDAO#2799]
	at org.hibernate.engine.jdbc.mutation.internal.ModelMutationHelper.identifiedResultsCheck(ModelMutationHelper.java:75)
	at org.hibernate.persister.entity.mutation.DeleteCoordinator.lambda$doStaticDelete$3(DeleteCoordinator.java:308)
	at org.hibernate.engine.jdbc.mutation.internal.ModelMutationHelper.checkResults(ModelMutationHelper.java:50)
	at org.hibernate.engine.jdbc.mutation.internal.AbstractMutationExecutor.performNonBatchedMutation(AbstractMutationExecutor.java:114)
	at org.hibernate.engine.jdbc.mutation.internal.MutationExecutorSingleNonBatched.performNonBatchedOperations(MutationExecutorSingleNonBatched.java:40)
	at org.hibernate.engine.jdbc.mutation.internal.AbstractMutationExecutor.execute(AbstractMutationExecutor.java:52)
	at org.hibernate.persister.entity.mutation.DeleteCoordinator.doStaticDelete(DeleteCoordinator.java:304)
	at org.hibernate.persister.entity.mutation.DeleteCoordinator.coordinateDelete(DeleteCoordinator.java:89)
	at org.hibernate.persister.entity.AbstractEntityPersister.delete(AbstractEntityPersister.java:2891)
	at org.hibernate.action.internal.EntityDeleteAction.execute(EntityDeleteAction.java:131)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:635)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:502)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:364)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:39)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1412)
	... 39 more
```

--- a/horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/SchemaServiceImpl.java
+++ b/horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/SchemaServiceImpl.java
@@ -474,8 +474,10 @@ public class SchemaServiceImpl implements SchemaService {
 
    private void doUpdateLabelForDelete(LabelDAO label) {
       em.createNativeQuery("DELETE FROM label_values WHERE label_id = ?1").setParameter(1, label.id).executeUpdate();
+      int schemaId = label.getSchemaId();
+      int labelId = label.id;
       label.delete();
-      emitLabelChanged(label);
+      emitLabelChanged(labelId, schemaId);
    }
 
    @PermitAll
@@ -659,7 +661,7 @@ public class SchemaServiceImpl implements SchemaService {
          label.schema = em.getReference(SchemaDAO.class, schemaId);
          checkSameName(label);
          label.persistAndFlush();
-         emitLabelChanged(label);
+         emitLabelChanged(label.id, schemaId);
       } else {
          LabelDAO existing = LabelDAO.findById(label.id);
          if (existing == null) {
@@ -690,20 +692,20 @@ public class SchemaServiceImpl implements SchemaService {
          existing.metrics = label.metrics;
          existing.persistAndFlush();
 
-         emitLabelChanged(existing);
+         emitLabelChanged(existing.id, existing.getSchemaId());
       }
       return label.id;
    }
 
-   private void emitLabelChanged(LabelDAO label) {
+   private void emitLabelChanged(int labelId, int schemaId) {
       try {
          List<Integer> datasetIds = session
                  .createNativeQuery("SELECT dataset_id from dataset_schemas WHERE schema_id = ?1")
-                 .setParameter(1, label.getSchemaId())
+                 .setParameter(1, schemaId)
                  .addScalar("dataset_id", StandardBasicTypes.INTEGER)
                  .getResultList();
          if (datasetIds == null || datasetIds.isEmpty() || datasetIds.get(0) < 1) {
-            log.debug("Could not extract datasetIds from dataset_schemas with schemaId="+label.getSchemaId());
+            log.debug("Could not extract datasetIds from dataset_schemas with schemaId="+schemaId);
             return;
          }
          int testId = session.createNativeQuery("SELECT testid from dataset WHERE id = ?1", Integer.class)
@@ -714,7 +716,7 @@ public class SchemaServiceImpl implements SchemaService {
          }
 
          for(var datasetId : datasetIds) {
-            Util.registerTxSynchronization(tm, txStatus -> mediator.queueDatasetEvents(new Dataset.EventNew(datasetId, testId, 0, label.id, true)));
+            Util.registerTxSynchronization(tm, txStatus -> mediator.queueDatasetEvents(new Dataset.EventNew(datasetId, testId, 0, labelId, true)));
          }
       }
       catch (NoResultException nre) {

